Bthome encryption downgrade

dafal opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This PR adds security protection against encryption downgrade attacks for BTHome devices. When a BTHome device that was previously configured with encryption suddenly starts broadcasting unencrypted data, this could indicate either:
1. The user intentionally disabled encryption on the device
2. A malicious actor attempting to spoof sensor data

**Implementation:**
- Uses the new `downgrade_detected` flag from bthome-ble v3.17.0 for reliable detection
- Blocks unencrypted payloads from devices with a configured bindkey (handled by library)
- Creates a persistent, fixable repair issue to alert the user
- Automatically clears the issue if encrypted broadcasts resume
- Provides a repair flow to remove the bindkey if encryption was intentionally disabled
This prevents silent acceptance of potentially spoofed unencrypted data from devices that should be encrypted.

**Why the library change was needed:**
The previous approach of checking `encryption_scheme == NONE` was unreliable because the library doesn't update the encryption scheme when blocking data. The new `downgrade_detected` flag explicitly signals when blocking occurs.

**Library dependency:**
- Requires bthome-ble >= 3.17.0
- Library PR: https://github.com/Bluetooth-Devices/bthome-ble/pull/309
- Changelog: Added `downgrade_detected` flag for integration detection

**Testing:**
- Tested with Shelly BLU H&T device
- Successfully detected encryption downgrades
- Repair issue created and cleared appropriately
- Fix flow validated (bindkey removal and entry reload)
- All 106 BTHome integration tests pass

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/160508
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Ernst79 commented:
Thanks for this PR, haven’t reviewed it yet, but there was also a recent change in the bthome-ble library. I’m not sure if you need this update, as it seems related?  https://github.com/Bluetooth-Devices/bthome-ble/pull/297. This PR is included in the 3.16.0 release of bthome-ble. If you need that, you probably also need to make a PR that bumps the bthome-ble version, as HA now uses 3.15.0

Ernst79 commented:
It would be good to change it to work with the new version, and adding the user experience like you have added now.

dafal commented:
Done, I hope it is ok but I did it using OpenCode and asked 3 models to confirm/control/challenge the changes. 🤞
Tested with a Shelly BLU H&T.

Ernst79 commented:
The bump of BTHome should be made in a different PR, as @thecode said. Just move the three files where you change the BTHome version into a separate PR

dafal committed:
feat(bthome): Add repair flow for encryption downgrade detection

Adds repair infrastructure to detect and handle BTHome devices that
downgrade from encrypted to unencrypted communication, which could
indicate a spoofing attempt or intentional encryption removal.

Uses the downgrade_detected flag introduced in bthome-ble 3.17.0
to reliably detect when a device configured with a bindkey starts
sending unencrypted data.

When downgrade is detected:
- Creates a repair issue alerting the user
- Blocks unencrypted data from being processed
- Provides fix flow to remove bindkey if change was intentional
- Automatically clears issue when encryption resumes

The repair flow gives users clear guidance on whether the downgrade
might be malicious or intentional, improving security awareness.

Depends on bthome-ble 3.17.0 bump (merged in #159681).

Tested with Shelly BLU H&T device successfully detecting encryption
downgrades and creating/clearing repair issues appropriately.

bdraco committed:
Merge branch 'dev' into bthome-encryption-downgrade

bdraco committed:
Merge branch 'dev' into bthome-encryption-downgrade

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into bthome-encryption-downgrade

bdraco committed:
Revert "Revert bthome-ble back to 3.16.0 to fix missing data (#160694)"

This reverts commit ce234d69a72a2df8a6826ab3308513d48d7a8890.

bdraco committed:
make sure repair is removed when entry is deleted

bdraco committed:
Merge remote-tracking branch 'dafal/bthome-encryption-downgrade' into bthome-encryption-downgrade

bdraco committed:
bool not needed, mypy can infer

bdraco committed:
clearer instructions

bdraco committed:
avoid accessing entry_data in tests

bdraco committed:
enums are singletons

bdraco committed:
rewrite test to use public api

bdraco committed:
caplog

bdraco committed:
reduce

bdraco committed:
better show intent

bdraco committed:
cleanup

bdraco commented on homeassistant/components/bthome/__init__.py:
```diff
@@ -36,6 +36,45 @@
+        "unencrypted data. This could be a spoofing attempt. "
+        "Data will be ignored until resolved",
+        entry.title,
+    )
```
After looking around, `LOGGER` around `async_create_issue` in the code base is rare so I tend to agree with removing it

bdraco committed:
Merge branch 'dev' into bthome-encryption-downgrade
